wfe2/wfe.go

```diff
@@ -2278,6 +2278,37 @@ func (wfe *WebFrontEndImpl) RenewalInfo(ctx context.Context, logEvent *web.Reque
 	logEvent.Extra["RequestedSerial"] = serial
 	beeline.AddFieldToTrace(ctx, "request.serial", serial)
 
+	// Check if the serial is part of an ongoing incident.
+	result, err := wfe.sa.IncidentsForSerial(ctx, &sapb.Serial{Serial: serial})
+	if err != nil && !errors.Is(err, berrors.NotFound) {
+		wfe.sendError(response, logEvent, probs.ServerInternal("Unable to check if the serial is impacted by an incident"), err)
+		return
+	}
+
+	if result != nil && len(result.Incidents) > 0 {
+		// Since IncidentsForSerial() only returns enabled incidents we can
+		// assume that this serial is impacted by an ongoing incident. Per the
+		// ARI spec clients should attempt to renew immediately if the suggested
+		// window is in the past.
+		oneHourAgo := wfe.clk.Now().Add(-1 * time.Hour)
+		ri := core.RenewalInfo{
+			SuggestedWindow: core.SuggestedWindow{
+				Start: oneHourAgo,
+				End:   oneHourAgo.Add(time.Minute * 30),
+			},
+		}
+
+		err = wfe.writeJsonResponse(response, logEvent, http.StatusOK, ri)
+		if err != nil {
+			wfe.sendError(response, logEvent, probs.ServerInternal("Error marshalling renewalInfo"), err)
+			return
+		}
+
+		pollPeriod := int(6 * time.Hour / time.Second)
+		response.Header().Set("Retry-After", fmt.Sprintf("%d", pollPeriod))
+		return
+	}
```

Comment:
All of this is basically duplicated from below. Can we come up with a streamlined way to just compute the renewal window using one of two mechanisms, and then only write the response and headers once?

Reply:
I've simplified this as much as I dare. I could break the writing of the JSON out into a helper for the sake of DRY but I feel it might hurt readability.